pam_ocra/pam_ocra.8
.\" Copyright (c) 2014, 2018 Stefan Grundmann
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"
.Dd April 9, 2018
.Dt PAM_OCRA 8
.Os
.Sh NAME
.Nm pam_ocra
.Nd RFC6287 OCRA: OATH Challenge-Response Algorithm PAM module
.Sh SYNOPSIS
.Op Ar service-name
.Ar module-type
.Ar control-flag
.Pa pam_ocra
.Op Ar options
.Sh DESCRIPTION
The
OCRA service module for PAM,
.Nm
provides functionality for only one PAM category:
authentication.
In terms of the
.Ar module-type
parameter, this is the
.Dq Li auth
feature.
It also provides null functions for the remaining module types.
.Ss OCRA Authentication Module
The OCRA authentication component
.Pq Fn pam_sm_authenticate
obtains OCRA credentials from the the per-user file
.Ar ~/.ocra .
If the
.Cm dir
parameter is set,
.Ar directory/USERNAME
will be used. It then
provides the user with an OCRA challenge and verifies the response.
.Pp
The following options may be passed to the authentication module:
.Bl -tag -width ".Cm disallow"
.It Cm dir Ns = Ns Ar directory
Specifies the additional directory to search for OCRA credentials.
.It Cm nodata Ns = Ns Ar action
Determines how the module handles the situation where there is no OCRA data
file associated with the user.  Use this option when some accounts use OCRA
authentication but other accounts do not.  The
.Ar action
value must be one of the following:
.Bl -tag -width ".So \  Sc (space)"
.It Sq Cm fail
In the absence of this option, or
if the
.Ar action
is set to
.Sq Cm fail
.Ns , an error message will be logged via
.Fn syslog
and PAM_AUTHINFO_UNAVAIL will be returned.  (But see
.Cm fake_prompt
.Ns , below.)
.It Sq Cm succeed
PAM_SUCCESS will be returned.
.It Sq Cm ignore
PAM_IGNORE will be returned.
.El
.Pp
Which option to use will depend on the
control flag used in PAM configuration file.
.It Cm fake_prompt Ns = Ns Ar suite_string
Use
.Ar suite_string
to generate fake challenges for users who do not have OCRA credentials.
Note that if this option is not set, no fake challenges will be generated
which can leak information to a hypothetical attacker about who uses OCRA
and who does not.
.Pp
If this option is specified, then the handling of the
.Cm nodata
option changes somewhat.  If the
.Cm nodata
option is absent, or the
.Ar action
is set to
.Sq Cm fail
.Ns , then the module will return PAM_AUTH_ERR instead of PAM_AUTHINFO_UNAVAIL.
.It Cm cmsg Ns = Ns Ar challenge_prompt Cm rmsg Ns = Ns Ar response_prompt
Change the challenge and/or the response prompts.  The
.Cm cmsg
option changes the challenge prompt, and the
.Cm rmsg
option changes the response prompt.  If the
.Cm cmsg
prompt is specified, a newline will be appended to it.  There will be no
newline appended to the
.Cm rmsg
prompt.  If spaces are included for either prompt, the prompt must be in
placed in double quotes.  For either prompt, the following formatting
directives may be used:
.Bl -tag -width ".So \  Sc (space)"
.It Sq Cm %c
Insert the challenge question.
.It Sq Cm %Nc
The challenge question with a spaces inserted after every N-th character
 (N>=1, N=<9).
.It Sq Cm %u
Insert a UTC timestamp in
.Cm ISO-8601
format. This information can be
useful when the OCRA suite string contains a time specification but the
clock on the system is unreliable.  Many SSH clients don't give visibility
to any system output prior to login, so this may be the only way to indicate
that a time discrepancy exists.  Note that the timezone abbreviation is
appended to the timestamp for readability purposes.  This timezone abbreviation
should be stripped off before parsing the timestamp.
.It Sq Cm %l
Insert a local-time timestamp in
.Cm ISO-8601
format.  (Which may still be UTC, depending on how the system is configured.)
In addition to the date and time, the timezone offset is appended to the local
timestamp.
Like its UTC counterpart, a readable timezone abbreviation is appended to the
timestamp.
.It Sq Cm %%
Insert a literal % character.
.El
.Pp
The default challenge prompt is "OCRA Challenge: %4c" and the default response
prompt is "OCRA Response: "
.El
.Sh FILES
.Bl -tag -width Ds -compact
.It Pa ~/.ocra
.El
OCRA credential file
.Sh NOTES
LinuxPAM does not handle quoted strings in pam module options.
When Linux PAM is uses instead of OpenPAM, options that contain spaces must be
surrounded by square brackets instead of quoting the option value.
.Sh EXAMPLES
Note that in the following examples, the pam_ocra.so entry in the PAM
configuration file is shown on multiple lines for readability purposes.  In the
actual configuration file, the module and its options must be on one line.
.Pp
A PAM config file with the following entries:
.Pp
.Bd -literal
auth required pam_unix.so no_warn null_ok
auth required pam_ocra.so \e
        nodata=succeed fake_prompt=OCRA-1:HOTP-SHA1-6:QN06-PSHA1
.Ed
.Pp
Would ask for both a normal login password and an OCRA response from all users.
If there is OCRA data associated with the user, then both authentication
methods must succeed.  A non-OCRA user only has to successfully enter the
normal login password.
.Pp
.Pp
A PAM config file with the following entries:
.Pp
.Bd -literal
auth requisite pam_unix.so no_warn null_ok
auth required pam_ocra.so nodata=fail
.Ed
.Pp
Would ask for a normal login password from all users, but only ask for an
OCRA response if the normal login succeeded
.Em and
there was OCRA data associated with the user.  For users without OCRA
data, the login would immediately fail.
.Pp
.Pp
For both of the above examples, the prompts would appear similar to the
following:
.Bd -literal
OCRA Challenge: 123456
OCRA Response:
.Ed
.Pp
.Pp
If the options included the following prompt changes:
.Bd -literal
cmsg="%u" rmsg="OTP Response to %c: "
.Pp
or in case LinuxPAM is used:
.Bd -literal
cmsg=%u [rmsg=OTP Response to %c: ]
.Ed
Then the prompts would look similar to:
.Bd -literal
2017-07-20T21:26:43Z UTC
OTP Response to 123456:
.Ed
.Pp
.Pp
Similarly if the options included the following prompt changes:
.Bd -literal
cmsg="%l - Challenge: %3c" rmsg="Response: "
.Pp
LinuxPAM version:
.Bd -literal
[cmsg=%l - Challenge: %3c] [rmsg=Response: ]
.Ed
.Pp
Then the prompts would look similar to:
.Bd -literal
2017-07-20T16:26:43-0500 CDT - Challenge: 123 456
Response:
.Ed
.Sh SEE ALSO
.Xr pam.conf 5 ,
.Xr pam 8 ,
.Xr ocra_tool 8
.Sh STANDARDS
.Bl -tag -offset indent -width 8n
.It Li RFC6287
.Em OCRA: OATH Challenge-Response Algorithm
.El
.Sh AUTHORS
The
.Nm
module and this manual page were developed by Stefan Grundmann


debian/changelog
libpam-ocra (1.5) unstable; urgency=low

  * Initial Release.

 -- Stefan Grundmann <sg2342@googlemail.com>  Fri, 06 Apr 2018 02:03:56 +0100


pam_ocra/pam_ocra.c
/*-
 * Copyright (c) 2014,2018 Stefan Grundmann
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <syslog.h>
#include <time.h>
#include <string.h>

#include <security/pam_modules.h>
#include <security/pam_appl.h>
#ifndef __linux__
#include <security/openpam.h>
#endif

#include "ocra.h"

#ifndef PAM_EXTERN
#define PAM_EXTERN
#endif

static int
adjust_return(const char *nodata, int ret)
{
	if (PAM_SUCCESS != ret && (PAM_AUTHINFO_UNAVAIL == ret ||
	    PAM_NO_MODULE_DATA == ret)) {
		/* Change the return code, if requested */
		if (NULL != nodata) {
			if (strcmp(nodata, "succeed") == 0) {
				ret = PAM_SUCCESS;
			} else if (strcmp(nodata, "ignore") == 0) {
				ret = PAM_IGNORE;
			} else if (strcmp(nodata, "fail") != 0) {
				syslog(LOG_ERR, "Unknown \"nodata\" value");
				ret = PAM_SERVICE_ERR;
			}
		}
		/*
		 * PAM_NO_MODULE_DATA is the result when a fake prompt is
		 * displayed.  If not handled above, treat it like an
		 * authentication failure.
		 */
		if (PAM_NO_MODULE_DATA == ret) {
			ret = PAM_AUTH_ERR;
		}
	}
	return ret;
}

static void
fmt_prompt(char *mbuf, int msize, const char *questions, const char *pmsg)
{
	char *mptr = mbuf;
	const char *pptr = pmsg;
	int mrsize = 0;
	time_t epoch_seconds;
	struct tm *now;
	char fmt[] = "%.Ns ";
	int N, i, j, ql;

	if (NULL == pmsg) {
		if (msize)
			*mptr = '\0';
		return;
	}
	msize--;			/* Ensure we always have room for
					 * trailing '\0' */

	ql = strlen(questions);

	while ((mrsize < msize) && '\0' != *pptr) {
		/* Copy over the first part of the string */
		while ((mrsize < msize) && '\0' != *pptr) {
			if ('%' != *pptr) {
				*mptr++ = *pptr++;
				mrsize++;
			} else {
				pptr++;
				break;
			}
		}

		/*
		 * Handle the conversion character.  If not understood,
		 * the '%' will be quietly dropped.
		 */
		switch (*pptr) {
		case '%':		/* Literal '%' */
			*mptr++ = '%';
			mrsize++;
			pptr++;
			break;

		case 'u':		/* UTC time */
			time(&epoch_seconds);
			now = gmtime(&epoch_seconds);
			strftime(mptr, msize - mrsize,
			    "%Y-%m-%dT%H:%M:%SZ %Z", now);
			mrsize = strlen(mbuf);
			mptr = &mbuf[mrsize];
			pptr++;
			break;

		case 'l':		/* Local time */
			time(&epoch_seconds);
			now = localtime(&epoch_seconds);
			strftime(mptr, msize - mrsize,
			    "%Y-%m-%dT%H:%M:%S%z %Z", now);
			mrsize = strlen(mbuf);
			mptr = &mbuf[mrsize];
			pptr++;
			break;

		case 'c':		/* Challenge question */
			snprintf(mptr, msize - mrsize,
			    "%s", questions);
			mrsize = strlen(mbuf);
			mptr = &mbuf[mrsize];
			pptr++;
			break;
		default:		/* Maybe %Nc */
			if ('0' < *pptr && '9' >= *pptr && 'c' == *(pptr + 1)) {
				fmt[2] = *pptr;	/* set precision in fmt to N */
				N = *pptr - '0';	/* int value */

				/* print with fmt */
				for (i = 0, j = ql % N ? ql / N : ql / N - 1;
				    0 < j; i += N, j--) {
					snprintf(mptr, msize - mrsize,
					    fmt, questions + i);
					mrsize = strlen(mbuf);
					mptr = &mbuf[mrsize];
				}

				/* print remaining */
				snprintf(mptr, msize - mrsize, "%s",
				    questions + i);
				mrsize = strlen(mbuf);
				mptr = &mbuf[mrsize];
				pptr += 2;
			}
			break;
		}
	}

	/* Terminate the prompt string */
	*mptr = '\0';
}

static void
make_prompt(char *buf, int bsize, const char *questions,
    const char *cmsg, const char *rmsg)
{
	char cbuf[512];
	char rbuf[512];

	/* Create the default prompt strings, if necessary */
	if (NULL == cmsg && NULL == rmsg) {
		cmsg = "OCRA Challenge: %4c";
		rmsg = "OCRA Response: ";
	}
	/* Generate each prompt */
	fmt_prompt(cbuf, sizeof(cbuf), questions, cmsg);
	fmt_prompt(rbuf, sizeof(rbuf), questions, rmsg);

	/* Concatenate them to the final prompt */
	if (NULL != cmsg && NULL != rmsg) {
		snprintf(buf, bsize, "%s\n%s", cbuf, rbuf);
	} else if (NULL != cmsg) {
		snprintf(buf, bsize, "%s\n", cbuf);
	} else {
		snprintf(buf, bsize, "%s", rbuf);
	}
}


static int
get_response(pam_handle_t *pamh, char *prompt, char **response)
{
	int ret;
	struct pam_message msg;
	const struct pam_message *msgp = &msg;
	const struct pam_conv *conv = NULL;
	struct pam_response *presponse = NULL;

	pam_get_item(pamh, PAM_CONV, (const void **)&conv);
	pam_set_item(pamh, PAM_AUTHTOK, NULL);

	msg.msg_style = PAM_PROMPT_ECHO_ON;
	msg.msg = prompt;

	ret = (*conv->conv) (1, &msgp, &presponse, conv->appdata_ptr);

	if (NULL != presponse) {
		if (PAM_SUCCESS == ret) {
			*response = presponse->resp;
			presponse->resp = NULL;
		}
	}
	return ret;
}

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags,
    int argc, const char *argv[])
{
	int qret;
	int ret;
	const char *dir = NULL;
	const char *fake_suite = NULL;
	const char *nodata = NULL;
	const char *cmsg = NULL;
	const char *rmsg = NULL;
	char *questions;
	const char *user;
	char *response = NULL;
	char fmt[512];

	(void)flags;

	/* Get options */
#ifdef __linux__
	while (argc--) {
		if (0 == strncmp(*argv, "fake_prompt=", 12))
			fake_suite = *argv + 12;
		else if (0 == strncmp(*argv, "dir=", 4))
			dir = *argv + 4;
		else if (0 == strncmp(*argv, "nodata=", 7))
			nodata = *argv + 7;
		else if (0 == strncmp(*argv, "cmsg=", 5))
			cmsg = *argv + 5;
		else if (0 == strncmp(*argv, "rmsg=", 5))
			rmsg = *argv + 5;
		argv++;
	}
#else
	(void)argc;
	(void)argv;

	fake_suite = openpam_get_option(pamh, "fake_prompt");
	dir = openpam_get_option(pamh, "dir");
	nodata = openpam_get_option(pamh, "nodata");
	cmsg = openpam_get_option(pamh, "cmsg");
	rmsg = openpam_get_option(pamh, "rmsg");
#endif
	pam_get_item(pamh, PAM_USER, (const void **)&user);

	openlog("pam_ocra", 0, LOG_AUTHPRIV);

	/*
	 * Generate the challenge "question".  If the user doesn't have any
	 * OCRA data, a fake challenge may be generated.
	 *
	 * Valid expected return codes are:
	 * 	PAM_SUCCESS          -	User has OCRA data and a valid challenge
	 * 				was generated.  A valid user response
	 * 				will be expected.
	 * 	PAM_AUTHINFO_UNAVAIL -	User does not have OCRA data and no fake
	 *		 		challenge was generated.  This result
	 *		 		may be modified based on the "nodata"
	 *		 		setting.
	 * 	PAM_NO_MODULE_DATA   -	User does not have OCRA data and a fake
	 * 				challenge was generated.  The ultimate
	 * 				return code after user input will be
	 * 				based on the "nodata" setting.
	 *
	 * 	Any other return code will be returned as-is.
	 */
	qret = challenge(dir, user, &questions, nodata, fake_suite);

	/* Only continue if there is a user prompt to display */
	if (PAM_SUCCESS != qret && PAM_NO_MODULE_DATA != qret) {
		ret = qret;
		goto end;
	}
	/* Generate the prompt */
	make_prompt(fmt, sizeof(fmt), questions, cmsg, rmsg);

	if (PAM_SUCCESS != (ret = get_response(pamh, fmt, &response)))
		goto end;

	if (PAM_SUCCESS != qret) {
		/*
		 * There was no OCRA data, so don't bother checking the
		 * response
		 */
		ret = qret;
	} else {
		ret = verify(dir, user, questions, response);
	}

	free(response);
end:
	closelog();
	return adjust_return(nodata, ret);
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char *argv[])
{
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;

	return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char *argv[])
{
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;

	return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc, const char *argv[])
{
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;

	return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char *argv[])
{
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;

	return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char *argv[])
{
	(void)pamh;
	(void)flags;
	(void)argc;
	(void)argv;

	return PAM_SUCCESS;
}

#ifdef PAM_MODULE_ENTRY
PAM_MODULE_ENTRY("pam_ocra");
#endif
